File: .opencode/memory/redis.js
// Minimal Redis client over a raw TCP socket. Redis's RESP protocol is simple enough that a
// dependency isn't worth it: we send RESP arrays and parse RESP replies. One persistent
// connection, promise-queued commands (Redis is single-threaded and replies in order).
// Used for the hot short-term tier + the write-stream. If Redis is down, callers treat it as a
// soft failure (memory still works via SQLite) -- this client never throws on connect loss, it
// resolves commands with a rejected promise the caller can swallow.

import net from "node:net"

function encode(args) {
  let s = `*${args.length}\r\n`
  for (const a of args) {
    const v = String(a)
    s += `$${Buffer.byteLength(v)}\r\n${v}\r\n`
  }
  return s
}

// Parse one RESP reply from buf starting at pos. Returns [value, nextPos] or null if incomplete.
function parse(buf, pos) {
  if (pos >= buf.length) return null
  const type = String.fromCharCode(buf[pos])
  const nl = buf.indexOf("\r\n", pos)
  if (nl === -1) return null
  const line = buf.toString("utf8", pos + 1, nl)
  const after = nl + 2
  switch (type) {
    case "+": return [line, after]
    case "-": return [new Error(line), after]
    case ":": return [Number(line), after]
    case "$": {
      const len = Number(line)
      if (len === -1) return [null, after]
      if (after + len + 2 > buf.length) return null
      return [buf.toString("utf8", after, after + len), after + len + 2]
    }
    case "*": {
      const n = Number(line)
      if (n === -1) return [null, after]
      const arr = []
      let p = after
      for (let i = 0; i < n; i++) {
        const r = parse(buf, p)
        if (!r) return null
        arr.push(r[0]); p = r[1]
      }
      return [arr, p]
    }
    default: return [line, after]
  }
}

export function createRedis(url = "redis://127.0.0.1:6379") {
  const u = new URL(url)
  const host = u.hostname || "127.0.0.1"
  const port = Number(u.port || 6379)

  let sock = null
  let buf = Buffer.alloc(0)
  const waiters = []          // FIFO of {resolve,reject} for pending replies
  let connected = false

  function flush() {
    let r
    while ((r = parse(buf, 0))) {
      buf = buf.subarray(r[1])
      const w = waiters.shift()
      if (!w) continue
      if (r[0] instanceof Error) w.reject(r[0]); else w.resolve(r[0])
    }
  }

  function connect() {
    return new Promise((resolve) => {
      sock = net.connect({ host, port }, () => { connected = true; resolve(true) })
      sock.on("data", (d) => { buf = Buffer.concat([buf, d]); flush() })
      sock.on("error", () => { connected = false })
      sock.on("close", () => {
        connected = false
        while (waiters.length) waiters.shift().reject(new Error("redis connection closed"))
      })
    })
  }

  async function cmd(...args) {
    if (!connected) await connect()
    if (!connected) throw new Error("redis unavailable")
    return new Promise((resolve, reject) => {
      waiters.push({ resolve, reject })
      sock.write(encode(args))
    })
  }

  return {
    connect,
    cmd,
    get connected() { return connected },
    close() { try { sock?.end() } catch { /* ignore */ } connected = false },
  }
}

File: core/log/redis.py
"""Minimal synchronous Redis client over a raw TCP socket.

RESP is simple enough that a dependency is not worth it: we send RESP arrays and parse
RESP replies. Used for the hot short-term tier + the write-stream. If Redis is down,
callers treat it as a soft failure (memory still works via SQLite) -- this client raises
RedisUnavailable on connect/command failure, which callers swallow.

Synchronous by design: the core is invoked per tool-call from a subprocess, so a blocking
socket is the right model. One connection, commands issued in order (Redis is
single-threaded and replies in order).
"""

from __future__ import annotations

import socket
from typing import Optional
from urllib.parse import urlparse


class RedisUnavailable(Exception):
    """Raised when Redis cannot be reached or a command fails. Callers swallow it."""


def _encode(args: tuple) -> bytes:
    out = [f"*{len(args)}\r\n".encode()]
    for a in args:
        v = str(a).encode()
        out.append(f"${len(v)}\r\n".encode())
        out.append(v)
        out.append(b"\r\n")
    return b"".join(out)


class Redis:
    """One persistent connection. Lazily connects on first command."""

    def __init__(self, url: str = "redis://127.0.0.1:6379", timeout: float = 1.5):
        u = urlparse(url)
        self.host = u.hostname or "127.0.0.1"
        self.port = u.port or 6379
        self.timeout = timeout
        self._sock: Optional[socket.socket] = None
        self._buf = b""

    def _connect(self) -> None:
        try:
            self._sock = socket.create_connection((self.host, self.port), timeout=self.timeout)
        except OSError as e:
            self._sock = None
            raise RedisUnavailable(str(e)) from e

    def cmd(self, *args):
        """Send one command and return its decoded reply. Raises RedisUnavailable on failure."""
        if self._sock is None:
            self._connect()
        try:
            self._sock.sendall(_encode(args))
            return self._read_reply()
        except (OSError, RedisUnavailable) as e:
            self.close()
            raise RedisUnavailable(str(e)) from e

    def _read_reply(self):
        while True:
            parsed = self._parse(self._buf, 0)
            if parsed is not None:
                value, nxt = parsed
                self._buf = self._buf[nxt:]
                if isinstance(value, _RespError):
                    raise RedisUnavailable(value.msg)
                return value
            chunk = self._sock.recv(65536)
            if not chunk:
                raise RedisUnavailable("connection closed")
            self._buf += chunk

    def _parse(self, buf: bytes, pos: int):
        if pos >= len(buf):
            return None
        t = buf[pos : pos + 1]
        nl = buf.find(b"\r\n", pos)
        if nl == -1:
            return None
        line = buf[pos + 1 : nl].decode()
        after = nl + 2
        if t == b"+":
            return (line, after)
        if t == b"-":
            return (_RespError(line), after)
        if t == b":":
            return (int(line), after)
        if t == b"$":
            n = int(line)
            if n == -1:
                return (None, after)
            if after + n + 2 > len(buf):
                return None
            return (buf[after : after + n].decode(), after + n + 2)
        if t == b"*":
            n = int(line)
            if n == -1:
                return (None, after)
            arr = []
            p = after
            for _ in range(n):
                r = self._parse(buf, p)
                if r is None:
                    return None
                arr.append(r[0])
                p = r[1]
            return (arr, p)
        return (line, after)

    def close(self) -> None:
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None
        self._buf = b""


class _RespError:
    def __init__(self, msg: str):
        self.msg = msg
